### ISSUE TITLE
`DateTime` arithmetic should not have wraparound behavior

### ISSUE THREAD

**Issue by Seelengrab** [JuliaLang/julia#50930]
For example:

```
julia> typemax(DateTime) + Year(146138512)
292277024-12-31T23:59:59

julia> typemax(DateTime) + Year(146138513)
-292277024-09-29T09:34:07.384

julia> typemin(DateTime) - Year(146138512)
-292277023-01-01T00:00:00

julia> typemin(DateTime) - Year(146138513)
292277025-04-03T14:25:51.616
```

These should error - it doesn't really make sense to have "wraparound" into the future/past.

**Comment by adienes** [comment]
while I agree with the premise, I'm not sure I would call this a correctness bug. The Dates docs are very clear that a `DateTime` is a relatively thin wrapper over `Int64` which is itself well-documented to have wraparound behavior

**Comment by JeffBezanson** [comment]
I don't really have an opinion on this issue but I do think the DateTime docs over-emphasize the implementation. It's fine to mention what the implementation is to give you a starting mental model, but I would not overly commit to the idea that a DateTime has to act like an Int64.

**Comment by jariji** [comment]
Thoughts

- (a) Who uses dates this extreme? Astronomers, I guess - are they using Dates.jl or AstroTime.jl? 
- (b) How bad is the performance on normal date operations? SaferIntegers.jl [says](https://jeffreysarnoff.github.io/SaferIntegers.jl/dev/benchmarks/) "Relative to the system Integer types, using the Safer Integer types has a time cost of 1.10x..1.25x..2.00x."
- (c) Do (a) and (b) actually matter? Julia shouldn't be having crazy results like this in Base - when you need to give up simplicity in exchange for performance, you should have to opt in. The only exception is raw numbers, because fast numbers are Julia's original selling point.

**Comment by Seelengrab** [comment]
Both python and Java don't have wraparound behavior (because it doesn't make sense for dates to wraparound into the future/past), and both are _extremely_ successful. I know that the default implementations for dates in those languages have lots of other issues, but those obviously haven't prevented people from building businesses on them.

I don't think performance is the be-all and end-all and I also don't think performance concerns should trump correctness in every case. This, in my opinion, is such a case.

**Comment by timholy** [comment]
@jariji the performance cost depends on the implementation. Checked arithmetic is not SIMD-compatible; with

```julia
function fastsum(list)
           s = zero(eltype(list))
           @turbo for i in eachindex(list)
               s += list[i]
           end
           return s
       end
```
(with `@turbo` from LoopVectorization), for `Int` vs `SafeInt` I get a factor of 5.5 on my fairly primitive CPU (AVX2), single-threaded. It would grow on a better CPU (e.g., approximately double again on an AVX512). So the claim made about SaferInteger's performance needs to be understood in the context in which it was intended (presumably "generic" use cases where branching and errors are possible).

I don't know whether performance really matters that much, though. @quinnj? I also don't know whether wraparound is anything more than a curiosity of toy problems; we have about 100MY of dynamic range with a resolution of 1ms. I'd guess that the relative imprecise encoding of short times might be more of a problem in practice than wraparound behavior, but honestly I don't know.

There's been a suggestion on slack that perhaps we could have `DateTime{T}` and then one could construct them from `Int` or `SafeInt` (or maybe even `Int128`) and get whichever behavior one wants. That's a breaking change, of course, but food for thought. Then the docs could focus on what it is/does and allow the numeric behavior to be specified by the type parameter.

**Comment by adienes** [comment]
> I don't know whether performance really matters that much, though.

I have a tough time envisioning a scenario where I would both
1. be ok with the unorthodox semantics of `DateTime`
2. require extreme, `Int`-like performance

From what I can gather, `DateTime` seems about as useful as an analog clock on the wall --- extremely useful! but not for high-performance or high-precision applications.

**Comment by Seelengrab** [comment]
The "discussion" on slack you're referencing is [here](https://julialang.slack.com/archives/C680MM7D4/p1692084791470079), which is me repeatedly trying to ask how wraparound is the correct semantic for a `DateTime` type. You refuse to answer that question, and keep going back to performance (or other completely irrelevant philosophical questions about overflow behavior in computer numbers in general), even after I pointed out, multiple times, that not everything is about performance and that I'm exclusively talking about `DateTime`. 

As mentioned above & on slack, using checked arithmetic for a `DateTime` type is not only common (Rust, C#, Java, Python all do it), but _pretty much expected_, because types representing points in time are usually only expected to increase. It is extremely common to have invariants such as

 * The user creation times in my DB only increase
 * Log entries have time points that only increase
 * All dates in context X are larger than Y

and having operations involving `DateTime` throw by default when these kinds of invariants are broken is very helpful to a programmer trying to build a program with consistent semantics. Without that, it's extremely likely that any operation involving `DateTime` needs to be wrapped, to check nothing funny is going on. Because ONCE something funny is going on (and that is not an IF, you need to plan for the non-happy path), checked arithmetic saves your butt by warning you of the funny business, instead of silently corrupting your data & breaking your invariants. And good look detecting & fixing that after the fact.

Exceptions should be thrown for an exceptional situation. Having a calculation travelling back in time is such an exceptional situation.

**Comment by timholy** [comment]
The fact that other languages seem to use checked arithmetic is probably the strongest argument for changing it. I took a brief look at Python, and they use checked, but python is slow so the overhead of checking may be irrelevant (the error thrown is about casting to a C-int, which makes me think Python is just wrapping a C library that uses 32-bit ints for its calculations). Rust I wasn't so sure about: I don't know Rust, but https://github.com/rust-datetime/datetime/blob/6c17632311da080ad174c66c2770c0373f3bce2b/src/instant.rs#L18-L21 and https://github.com/rust-datetime/datetime/blob/6c17632311da080ad174c66c2770c0373f3bce2b/src/instant.rs#L67-L89 seem, to a completely naive reader who's not even sure that `i64` is what I think it is, to be unchecked. Can you clarify, @Seelengrab?

If there isn't consensus among other (fast) languages, then it's worth remembering that this issue doesn't seem to have come from a real-world bug: it came from someone playing around with `typemin`/`typemax` and observing that bad things can happen. We get lots of such bug reports, and we treat them individually: sometimes it's a 4-alarm fire that we're grateful to learn about and promptly fix, and other times we close the issue with a shrug.

Moreover, we're talking about changing behavior, and so it's not wrong to think about it carefully. We could make the change and no one complains (maybe even some packages fix bugs...), that would be a great outcome. Or we could release Julia 1.11, and a month later get a complaint from a healthcare informatics company saying that their database query on cancer remission now runs 8% more slowly, which for them translates into $2M a year in extra server costs. We could say "but aren't you happy with the extra safety?" and they might respond that this form of theoretical safety is irrelevant to any actual real-world application that anyone would ever write. To me this doesn't seem a likely outcome, but it's not not wrong to pause to consider it.

It would be awesome if someone for whom datetime is their bread-and-butter could chime in---I just don't know about the realities to know, and I'm unsure of whether anyone else participating in this conversation does either. Until then, I'll emphasize that making it parametric might be the best of all worlds: we could have `DateTime` default to `DateTime{SafeInt64}` but then tell the healthcare dev team that they can change their database to use `DateTime{Int64}` and restore their former performance.

**Comment by adienes** [comment]
> [Rust's datetimes] seem [...] to be unchecked

Just a quick note that in Rust, ALL signed integer under/overflow will "panic" when run in debug mode (although maybe their datetimes also wrap in release mode)

**Comment by nalimilan** [comment]
> The "discussion" on slack you're referencing is [here](https://julialang.slack.com/archives/C680MM7D4/p1692084791470079), which is me repeatedly trying to ask how wraparound is the correct semantic for a `DateTime` type. You refuse to answer that question, and keep going back to performance (or other completely irrelevant philosophical questions about overflow behavior in computer numbers in general), even after I pointed out, multiple times, that not everything is about performance and that I'm exclusively talking about the semantics of `DateTime`.

> Should I stop reporting such issues, because I haven't found them "working on real code" (who says I haven't? that's just speculation on your part)?

@Seelengrab Can we avoid making this a personal issue? @timholy gave a balanced opinion, and he didn't even say we shouldn't change the behavior of `DateTime`. This is not a fight, and anyway other people will have to chime in before making a decision. You're not going to convince them by attacking Tim.

There is clearly a tradeoff between performance and safety here: there's no "right" and "wrong" solution. This behavior has been implemented for years, so we can take the time to have an in-depth discussion about whether it should be changed. (Personally I would like to hear the feedback from intensive users of `DateTime`.)

**Comment by Seelengrab** [comment]
I'm not trying to attack Tim, I've just repeatedly asked the question "Why is wraparound semantically correct?" and the only answer I got is "because `DateTime` is like a fixed point number", which is not a justification at all. Nothing about fixed-point-ness prescribes wraparound behavior. The other responses I got were in the same vain as

![screenshot_2023-08-16_09-14-19_398982914](https://github.com/JuliaLang/julia/assets/11753998/e249b9b9-d0d4-445b-b901-71f9a9b46eea)

([link](https://julialang.slack.com/archives/C680MM7D4/p1692169556550469?thread_ts=1692084791.470079&cid=C680MM7D4)) which I thought we as a community were long past (and hence my continued frustration with trying to argue for better semantics of this type - the changes I'm suggesting end up making code using this type more correct by default, preventing possible bugs by design). "correctness is not an issue" and "it's only incorrect if it breaks some documented behavior" fly in the face of our stance that we should (and MUST) do better here. Following that train of thought to its conclusion would mean not documenting anything, because then at least our code would always be correct! That's obviously absurd. You all remember how much of a PR disaster the last correctness discussion was, and I'd rather not repeat that.

I've given plenty of situations where the current behavior leads to wrong results without erroring, which is exactly what the `correctness bug` label describes. I've yet to hear a good _semantic_ (not "current implementation" or the fact that it can be viewed as a fixed point number) argument for why `DateTime` in particular _should_ wrap around (irrespective of the way its implemented) - and my continued stance is that anyone relying on the wraparound behavior has very broken code & subtle bugs.

**Comment by nalimilan** [comment]
In this issue the only debate I can see in comments is whether the performance loss due to checking for overflow would be acceptable or not. This is a classic question in Julia, see e.g. https://github.com/JuliaStats/Statistics.jl/issues/22.

**Comment by Seelengrab** [comment]
Yes - as I said above, I thought we as a community had learned by now that sometimes, performance has to take a backseat to not silently giving a bad/wrong result. That's the argument I've been making since the very beginning of this issue, and I've been making it very clear that sacrificing performance here is perfectly acceptable. We already check quite a lot of stuff in the `DateTime`/`Date` constructors through a throwing path, so I'd be very surprised if this has any additional performance impact at all.

Note also that in the issue you linked, the overflow was ultimately fixed due to it being buggy behavior.

**Comment by Seelengrab** [comment]
Let me put it this way: Erroring instead of wrapping around can prevent bugs like these:

 * https://en.wikipedia.org/wiki/Year_2038_problem
 * https://en.wikipedia.org/wiki/Year_2000_problem

and wikipedia even maintains a list of bugs related to this kind of overflow issue:

https://en.wikipedia.org/wiki/Time_formatting_and_storage_bugs

So I really don't think this is just a theoretical concern or "not a real problem".

**Comment by elextr** [comment]
> Until then, I'll emphasize that making it parametric might be the best of all worlds: we could have DateTime default to DateTime{SafeInt64} but then tell the healthcare dev team that they can change their database to use DateTime{Int64} and restore their former performance.

Why is no-one saying "yay, have our cake and eat it too", safe by default, and speed when there is the need?

**Comment by Seelengrab** [comment]
As mentioned above, the results of `DateTime` arithmetic are already checked. They need to be, in order to be able to construct a valid `DateTime` object in the first place.

This is not just an issue of the underlying/wrapped integer overflowing or requiring checked integer arithmetic, but an issue of `DateTime` arithmetic needing to be checked in its own context. The implementation as a wrapper of an `Int64` is incidental.

**Comment by timholy** [comment]
The ["full" constructors](https://github.com/JuliaLang/julia/blob/ac607dc5184e66183cf7375152c5503f2e57d78f/stdlib/Dates/src/types.jl#L218-L230) check, but you can construct without checking if you go [directly from `Milliseconds`](https://github.com/JuliaLang/julia/blob/ac607dc5184e66183cf7375152c5503f2e57d78f/stdlib/Dates/src/types.jl#L150) and [that's what we use in some arithmetic](https://github.com/JuliaLang/julia/blob/ac607dc5184e66183cf7375152c5503f2e57d78f/stdlib/Dates/src/arithmetic.jl#L82).

**Comment by Seelengrab** [comment]
> The ["full" constructors](https://github.com/JuliaLang/julia/blob/ac607dc5184e66183cf7375152c5503f2e57d78f/stdlib/Dates/src/types.jl#L218-L230) check

Well.. yes and no - they don't check the years (see https://github.com/JuliaLang/julia/issues/50981, for reference) and the `toms` function also doesn't check for overflow

https://github.com/JuliaLang/julia/blob/74ce6cf070a2a04e836c3e5a2211228a3ac978ef/stdlib/Dates/src/periods.jl#L445C1-L445C1

("documented" in a comment :upside_down_face: I guess this means that the constructor from an `UTM` is considered internal?)

**Comment by Seelengrab** [comment]
> Especially given that there is a technical solution that achieves both.

Could you expand on that? I don't follow how

https://github.com/JuliaLang/julia/blob/ac607dc5184e66183cf7375152c5503f2e57d78f/stdlib/Dates/src/arithmetic.jl#L82

can be made safe & correct without checked arithmetic. Increasing the precision of `DateTime` is not enough, because you can still overflow on bad input, leading to the undesired wraparound.

**Comment by timholy** [comment]
`DateTime{SafeInt64}`: checked; `DateTime{Int64}`: fast.

**Comment by Seelengrab** [comment]
That's a breaking change though. My argument in this issue is that the current behavior is buggy, and making it error should not be breaking.

**Comment by timholy** [comment]
Sure. Can we stop going in cycles? The bottom line is this: we can make it check, at the cost of some runtime performance. We're a little unsure of the tradeoff between risk of overflow and risk that performance matters. There are more ambitious solutions that allow users to make that choice for themselves. These more ambitious solutions may or may not be breaking (one can make `DateTime(args...) = DateTime{SafeInt64}(args...)` if you like...), but it would take someone sitting down and implementing it to be sure.

**Comment by Seelengrab** [comment]
I'm not sure what you're telling me - is it alright with you if I open a PR introducing the checked arithmetic, even without providing an "escape hatch"?

**Comment by timholy** [comment]
Certainly!

**Comment by bvdmitri** [comment]
If performance is a concern, then making it `DateTime{T}` is a bad option IMO, because it will make the existing code type-unstable and as a consequence less performant. Typing `DateTime{Int}` everywhere is annoying and (for no obvious reason) leaks internal implementation detail. It would also raise tons of other issues such as should `DateTime{Int8}` work and how to map into a date?

**Comment by timholy** [comment]
> because it will make the existing code type-unstable

```julia
julia> struct Foo{T}
           x::T
       end

julia> Foo(x) = Foo{Int}(x)
Foo

julia> @code_typed Foo(0xff)
CodeInfo(
1 ─ %1 = Core.zext_int(Core.Int64, x)::Int64
│   %2 = %new(Foo{Int64}, %1)::Foo{Int64}
└──      return %2
) => Foo{Int64}
```

**Comment by bvdmitri** [comment]
What I've meant is that if you have something like 

```
struct MyStruct
    datetime::DateTime
end
```
or 
```
Vector{DateTime}
```
Making it `DateTime` a parametric type would make the related code less performant because the `DateTime` would be an abstract type.

Doing
```
const DateTime = InnerDateTime{SafeInt64}
```
certainly solves this particular issue but not `InnerDateTime{Int8}`. 

I personally just wanted to give some external opinion. I don't use `DateTime` too much, but if I would I wouldn't expect it to have top-class performance characteristics. I actually expect it to be somewhat slowish, because my mental model of dates and times in programming assumes a lot of weird edge cases/extra checks/different time zones/etc. Maybe it is not applicable to `DateTime`, but these are my personal assumptions. If I would need top-performance I would probably look for a external top-performance `DateTime` library in the first place of instead using the `Base` version. On the other hand I also agree with the fact that meeting overflow behaviour in any practical use case sounds unrealistic if you have such a large time span. 

Also, the whole issue to me seems more like a confusing usage of `+` and `-` in the first place. A date is conceptually **not** a number, from a user perspective it would make more sense to define operations like `diff` between `DateTime` objects instead of using `-` directly. These `diff`-like operations could have `strict = false` keyword argument, for example, which wouldn't check for overflow and have better performance. I'm not sure however are other libraries or other programming languages also using `+` and `-`.

**Comment by timholy** [comment]
> Making it DateTime a parametric type would make the related code less performant because the DateTime would be an abstract type.

Excellent point! (There and elsewhere.)

Most of the libraries I've checked (which is not many) do seem to support `+` and `-`, so we're not unique in that regard.

**Comment by alecloudenback** [comment]
I think finance is an area where high-performance dates may matter. @felipenoris or @simonbyrne may be able to provide some insight here.

**Comment by felipenoris** [comment]
In the Finance perspective, I would argue that unchecked arithmetic makes more sense given the performance, and also the fact that finance computations goes just a few years ahead (maybe 100 years at most), so the wraparound behavior is not an issue.
If someone really needs it, maybe a method safe_add should do it, or maybe a new type SafeDateTime, instead of using a parametric type to avoid a breaking change.

**Comment by alecloudenback** [comment]
Knowing very little about the implementation, could the whole Dates package as exists now be spiked out as an FastDates package while the Dates embedded within Julia gets overflow checked?

**Comment by PallHaraldsson** [comment]
>typemax(DateTime) + Year(146138513)

I think nobody would actually do that, and I'm thinking, shouldn't we rather lower typemax of DateTime and/or for Year, and up the correspending typemin.

**Comment by adienes** [comment]
>  For high-frequency trading I could easily imagine that it could matter
> except that we probably don't even have the necessary resolution in DateTime.

Indeed.  also the fact that working with unix times is pretty unergonomic (why does `datetime2unix` return a `Float64` ???)

I would strongly recommend entirely forgetting about the use-case of HFT as a justification for performance, since no such users will choose `DateTime` for timestamps anyway

**Comment by Seelengrab** [comment]
> Not for anything less than a month:

I think @simonbyrne was referring to the fact that lots of _user_ code involving calculations on `DateTime` are using branches, precisely because they're very often used in cutoff dates and the like. I don't think he meant that `DateTime` calculations themselves are necessarily branchy.

**Comment by timholy** [comment]
True, but if you offset by a `Month` or `Year`, then DateTime internals are branchy too (no escape, then).

I do think we've waited long enough for someone to tell us that performance matters, without a convincing case yet, that I certainly wouldn't personally stand in the way of making it checked. Just be aware that if complaints roll in after a release, there's always a chance it would need to be reverted.

**Comment by LilithHafner** [comment]
Removing the label since this is not "likely to lead to incorrect results in user code without throwing". If it's documented to check for overflow (I don't think it is) then it's a bug.

Whether or not it's documented to check for overflow it is probably acceptable to add that check as a minor change if we decide the runtime cost is worth it.

**Comment by jariji** [comment]
> If it's documented to check for overflow (I don't think it is) then it's a bug.

I strongly disagree that the definition of a bug is circumscribed by what is in the (generally very limited) docstring. A datetime type should accord with reasonable axioms for a datetime concept, which is much more expansive. 

In particular, time does not wrap; neither should `DateTime` be expected to, without _explicit_ documentation to the contrary.

**Comment by simonbyrne** [comment]
Sure, it's a bug, but at the point at which becomes an issue (in 150 million years), changes in the Earth's rotation mean that our current calendar will be largely useless. If someone wants to fix it, I will be happy to review, but it's a pretty low priority

**Comment by PallHaraldsson** [comment]
Why is it a "bug" "for sure"? If it will never practically be triggered in millions of years?

@Seelengrab
> we should (and MUST) do better here

You feel strongly about this, and I don't, so would you be ok with a) just documenting possible wraparound: "will in practice happen in millions of years", or b) insist on checked? These seem exclusionary, but we could add checked for 1.11 (or 1.12) and NOT promise that it will happen (explicitly state it's experimental, that "safety" may be dropped at any time, could it even be a global setting?), in the unlikely case we get a regression report, and disable it in a quick point release:

>we could release Julia 1.11, and a month later get a complaint from a healthcare informatics company saying that their database query on cancer remission now runs 8% more slowly, which for them translates into $2M a year in extra server costs. We could say "but aren't you happy with the extra safety?"

**Comment by adienes** [comment]
> never practically be triggered in millions of years?

this is only true if your code is correct, which is usually not how bugs arise

I think it's quite surprising that this is at all controversial. people seem to be using financial applications as a strawman to prevent the implementation of checked arithmetic, but as someone who works in HFT myself, I promise you that _nobody_ (profitable) will _ever_ use `Dates` or `DateTime` for this kind of high-granularity low-latency application, whether or not the arithmetic is checked.

**Comment by alecloudenback** [comment]
@adienes I don’t think finance is used as a straw man. I just raised it as an area where high performance dates might matter and suggested that we consider gathering thoughts from those that might have experience in that area. Do you mind providing more context about how you see this issue and what you would do in practice to either accept the limitations of the implementations or what you would look for in a Dates implementation?

**Comment by PallHaraldsson** [comment]
>this is only true if your code is correct, which is usually not how bugs arise

I mean yes, this "bug" only triggers if you do "+ Year(146138513)" (or add more or subtract similar ridiculousness; there's no way to multiply), once;  or add up to that much in a loop in many steps. If you have sensible input validation limiting to YYYY then that's fairly many loop counts. And this is assuming you have typemax to begin with, so running that loop would take years. If you start with the year 9999, and add that repeatedly, then it seems to be 43845 loop counts, 213861 if you start with 2050. Then you get negative number that will be obvious, if you're showing them (or your database will fail to insert those), until you in a lot more iterations get to positive years again.

As I said I can go either way, just closing this bug (by documenting), or add a check, but I wouldn't commit to that check right away. If we do it would be a breaking change to undo it in case slow.

I would support throwing on constructing Year larger than 10000 to mitigate this, by default, for those who really need and want that it could be Year(1000; allow_huge=true) and they would take the consequences or know to opt into another type. Those are people likely needing a different type for astro anyway.

**Comment by adienes** [comment]
>  providing more context about how you see this issue and what you would do in practice to either accept the limitations of the implementation or what you would look for in a Dates implementation?

I genuinely don't think this would be valuable, because it's just not the right tool. `DateTime` will never ever ever be used for high-frequency trading timestamps. it's just not even worth considering that use case.

`Dates` should instead continue to play to its strengths which are convenience and ergonomics, for which imo checked arithmetic is a no brainer to add

**Comment by PallHaraldsson** [comment]
We accept far more dangerous unchecked integer arithmetic all the time, but worry here where everything but least dangerous operations are impossible. You can even add to typemax without overflow! People don't complain much about the default modular Int64, except for ^ since power is very dangerous and that problem should be avoided. Multiplying (and shifting) is also dangerous, and neither apply here. The only possibility of overflow are the least dangerous operations +  and -, which grow very slowly.

As I said, I can accept checked, it just seems odd to worry about this. It hasn't been a problem, seems very much made up "problem", while we really should worry about other things e.g. ^.

**Comment by PallHaraldsson** [comment]
I thought and still think only integers used (so what operation did you have in mind?):
```
julia> dump(ans)
DateTime
  instant: Dates.UTInstant{Millisecond}
    periods: Millisecond
      value: Int64 63713520000000
```

"converting to/from unix epoch which is a very hard requirement", seems easy, if not done, just do it? Could be functionality in or out of the stdlib. [I'm not sure if this has to do with leap-seconds that the ISO standard and DateTime do not support, but I doubt it, and not sure what they mean if anything for HFT.] I'm unclear about rounding behavior or need for, unless to convert from lower precision, e.g. micro or nanoseconds. I can see you want such for HFT, possibly, at least better than "Millisecond", which is all you get, not microseconds currently.

Converting from e.g. microseconds seemingly shouldn't do rounding it that's actually done, rather ceil. 

[I meant no disrespect to anyone (here), nor call anyone incompetent. I meant competent programmers would not overflow, nor do I think others really in this case, but I can't rule that out, nor for either, ok. And being incompetent isn't a slight, anyone is/a beginner at some point, so I just meant as a description. Neither did I mean they deserve "bugs", or traps. I'm really just trying to foresee how overflows might happen for anyone, trying to come up with a non-contrieved case. Only competent people, like the very competent @Seelengrab think of this, for the benefit of others.]

HFT is maybe a distraction here, I'm not trying to shoehorn it in, nor did I bring it up first. You mentioned it, as it wouldn't work, I was curious what would, or not, work, since adding and subtracting to/from DateTime is literally one assembly instruction. I think you may have something else in mind regarding extra overhead. I  confirmed that with a test I did `@code_native D + ms` (plus 4 trivial that should be eliminated when inlined; adding seconds rather than milliseconds does `add imulq	$1000, (%rsi), %rax`.

Adding nano (or micro) seconds only add a trivial about of very fast code:
```
movabsq	$4835703278458516699, %rax      # imm = 0x431BDE82D7B634DB
; │┌ @ /cache/build/default-amdci5-5/julialang/julia-release-1-dot-9/usr/share/julia/stdlib/v1.9/Dates/src/periods.jl:446 within `toms`
; ││┌ @ int.jl:295 within `div`
	imulq	(%rsi)
	movq	%rdx, %rax
	shrq	$63, %rax
	sarq	$18, %rdx
```

**Comment by adienes** [comment]
>  I meant competent programmers would not overflow,

I just don't think this is true. I have seen very competent engineers make goofs

everybody does it


> I thought and still think only integers used (so what operation did you have in mind?)

ok, you are right kinda, but in another issue it was decided that despite the underlying object being integer, the behavior is "fixed-precision-like" with respect to rounding etc.

>  seems easy, if not done, just do it?

ok, how? anything more complicated than a performant `epoch(t)` is a nonstarter

> Converting from e.g. microseconds seemingly shouldn't do rounding it that's actually done, rather ceil.
agree, but ship has sailed. there was a lot of discussion and then PR merged
